api/src/main/java/org/jenkinsci/plugins/workflow/graph/FlowGraphWalker.java

```diff
@@ -53,4 +56,59 @@ public FlowNode next() {
         }
         return null;
     }
+
+    @Override
+    public Iterator<FlowNode> iterator() {
+        return new FlowGraphWalkerIterator();
+    }
+
+    private class FlowGraphWalkerIterator implements Iterator<FlowNode> {
+
+        private FlowNode next;
+        private int expectedCount;
+
+        public FlowGraphWalkerIterator() {
+            expectedCount = q.size();
+            next = getNext();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return next != null;
+        }
+
+        @Override
+        public FlowNode next() {
+            FlowNode temp = next;
+            next = getNext();
+            return temp;
+        }
+
+        @Override
+        public void remove() {
+            getNext();
+        }
+
+        private FlowNode getNext() {
+            checkForComodification();
+            while (!q.isEmpty()) {
+                FlowNode n = q.pop();
+                if (!visited.add(n)) {
+                    continue;
+                }
+
+                List<FlowNode> parents = n.getParents();
+                addHeads(parents);
+                expectedCount = q.size();
+                return n;
+            }
+            return null;
+        }
+
+        private final void checkForComodification() {
+            if (q.size() != expectedCount) {
+                throw new ConcurrentModificationException();
```

Comment:
I've not found any call to deprecated ```next()``` in the code which could produce that issue. Anyway, if we want to comply with the ```Iterable``` contract we have to check for concurrent modifications. The only way to modify a collection while iterating is by using the ```Iterator.remove()``` method (which is not supported here neither).